internal/runners/ppm/convert.go

```diff
@@ -0,0 +1,230 @@
+package ppm
+
+import (
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+
+	"github.com/ActiveState/cli/internal/analytics"
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/failures"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/skratchdot/open-golang/open"
+)
+
+// convertAnswerCreate is the answer that the user can choose if they accept to create a virtual environment.  It is re-used at several places.
+var convertAnswerCreate string = locale.Tl("ppm_convert_answer_create", "Create Virtual Runtime Environment")
+
+// analyticsEventFunc is used to send analytics event
+type analyticsEventFunc func(string, string, string)
+
+// surveySelectFunc displays a menu with options that the user can select
+type surveySelectFunc func(message string, choices []string, defaultResponse string) (string, *failures.Failure)
+
+type conversionFlow struct {
+	survey   surveySelectFunc
+	out      output.Outputer
+	openURI  func(string) error
+	visitedC chan string
+	once     sync.Once
+}
+
+// StartConversionFlowIfNecessary checks if the user is in a project directory.
+// If not, they are asked to create a project, and (in a wizard-kind-of way) informed about the consequences.
+func (p *Ppm) StartConversionFlowIfNecessary() error {
+	// start conversion flow only if we cannot find a project file
+	if p.project == nil {
+		c := make(chan os.Signal, 1)
+		defer close(c)
+		signal.Notify(c, os.Interrupt)
+		defer signal.Stop(c)
+		cf := newConversionFlow(p.prompt.Select, p.out, open.Run)
+		defer cf.Close()
+
+		r, err := cf.run(c, analytics.EventWithLabel, func() { os.Exit(1) })
+
+		if r == accepted {
+			cf.createVirtualEnv()
+		}
+		return err
+	}
+	return nil
+}
+
+func newConversionFlow(survey surveySelectFunc, out output.Outputer, openURI func(string) error) *conversionFlow {
+	// Note: the buffer length needs to be big enough to hold the maximum number of visited items
+	visitedC := make(chan string, 20)
```

Comment:
This is trying to be too smart for its own good. Keep it dumb and easy to digest. I'd much rather see some redundant analytic event calls than layers of abstraction and channel magic so we can avoid a bit of redundancy.

Reply:
It relates to the answer above.  But on second thought (considering what you sent on slack about the interrupts in the `survey` library), I could probably just handle interrupts through returned error codes from the survey library.  Then I would not need go functions and all this channel stuff that took me so long to get right. :scream: 

Yes, I will do that! :)  It will look much better afterwards!